File: basic/src/main/kotlin/com/explore/truth/universe/binarysearch/gzz/BinarySearch.java
package com.explore.truth.universe.binarysearch.gzz;

public class BinarySearch {
    public static int search(int[] intArray, int target) {
        int result = -1;
        int start = 0;
        int end = intArray.length - 1;

        while (true) {
            if (target > intArray[start] && target < intArray[end]) {
                int mid = (start + end) / 2;
                if (mid <= start || mid >= end) {
                    break;
                }

                if (target < intArray[mid]) {
                    end = mid;
                } else if (target > intArray[mid]) {
                    start = mid;
                } else {
                    result = mid;
                    break;
                }
            } else if (target == intArray[start]) {
                result = start;
                break;
            } else if (target == intArray[end]) {
                result = end;
                break;
            } else {
                break;
            }

        }

        return result;
    }
}

File: basic/src/main/kotlin/com/explore/truth/universe/binarysearch/cooper/BinarySearch.kt
package com.explore.truth.universe.binarysearch.cooper

/**
 * Created by cooper
 * 2020/5/29.
 * Email: 1239604859@qq.com
 */

/**
 * @return Int: 如果存在的话范围source对应的位置, 不存在返回 -1
 */
fun binarySearch(source: IntArray, target: Int): Int {
    var start = 0
    var end = source.size - 1
    while (start <= end) {
        val mid = (start + end).shr(1)
        val midValue = source[mid]
        when {
            midValue < target -> {
                start = mid + 1
            }
            midValue > target -> {
                end = mid - 1
            }
            else -> {
                return mid
            }
        }
    }
    return -1
}
